### DISCUSSION TITLE
[JS 스터디] DOM

### DISCUSSION THREAD

**Discussion by hochan222** [transcendence42/javascript-archive#5] [JS스터디]
주제 키워드는

✨DOM✨

입니다.

#### ☘️  하위 주제 키워드

- querySelector
- getElementById
- EventListener

⏳스레드에 각 페어당 소요시간 적어주세요!

(ex) holee 90분 소요

참고링크
- https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector
- https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById
- https://developer.mozilla.org/ko/docs/Web/API/EventListener

**Comment by hochan222** [comment]
공유하고싶은 자료 & 개인적으로 정리한 블로그 등이 있다면 여기👇👇👇에 댓글로 공유해주세요 😆

**Reply by hochan222** [reply]
- DOM
  - https://developer.mozilla.org/ko/docs/Web/API/Document_Object_Model/Introduction
- querySelector, getElementById
  - https://whatabouthtml.com/difference-between-getelementbyid-and-queryselector-180

**Reply by yechoi42** [reply]
- 클래스와 기본 문법
   - https://ko.javascript.info/class

**Reply by yechoi42** [reply]
(꼭 관련있다고 할 순 없지만...)
- template literal
   -  https://joshua1988.github.io/es6-online-book/template-literal.html#%EC%97%AC%EB%9F%AC-%EC%A4%84%EC%97%90-%EA%B1%B8%EC%B3%90-%EB%AC%B8%EC%9E%90%EC%97%B4-%EC%84%A0%EC%96%B8%ED%95%98%EA%B8%B0

**Reply by jwon42** [reply]
- DOM은 정확히 무엇일까? (번역)
  https://wit.nts-corp.com/2019/02/14/5522
- What, exactly, is the DOM? (원문)
  https://bitsofco.de/what-exactly-is-the-dom/
- 브라우저는 어떻게 동작하는가?
  https://d2.naver.com/helloworld/59361

**Comment by hochan222** [comment]
@yhshin0 
holee 40분 소요

## 1. DOM 은 무엇인가요?

DOM 은 동일한 문서를 표현하고, 저장하고, 조작하는 방법을 제공한다. DOM 은 웹 페이지의 객체 지향 표현이며, 자바스크립트와 같은 스크립팅 언어를 이용해 DOM 을 수정할 수 있다.

문서 객체 모델(The Document Object Model, 이하 DOM) 은 HTML, XML 문서의 프로그래밍 interface 이다. DOM은 문서의 구조화된 표현(structured representation)을 제공하며 프로그래밍 언어가 DOM 구조에 접근할 수 있는 방법을 제공하여 그들이 문서 구조, 스타일, 내용 등을 변경할 수 있게 돕는다.

```js
var paragraphs = document.getElementsByTagName("P");
```
DOM 은 프로그래밍 언어는 아니며, DOM 은 프로그래밍 언어와 독립적으로 디자인되었다. DOM 의 구현은 어떠한 언어에서도 가능하다.

```python
# Python DOM example
import xml.dom.minidom as m
doc = m.parse("C:\\Projects\\Py\\chap1.xml");
doc.nodeName # DOM property of document object;
p_list = doc.getElementsByTagName("para");
```

각각의 브라우저는 자신만의 방법으로 DOM 구현해서 브라우저마다 구현 사항이 다르다.

아래는 웹 페이지, XML 페이지 스크립팅에서 DOM 을 사용하는 공통적인 API 들의 간략한 목록이다. 

- document.getElementById(id)
- document.getElementsByTagName (en-US)(name)
- document.createElement(name)
- parentNode.appendChild (en-US)(node)
- element.innerHTML (en-US)
- element.style.left
- element.setAttribute (en-US)
- element.getAttribute
- element.addEventListener (en-US)
- window.content (en-US)
- window.onload (en-US)
- window.dump (en-US)
- window.scrollTo (en-US)

## 2. querySelector와 getElementById 차이점은 무엇인가요?

차이점은 getElementById가 querySelector보다 성능/지원 측면에서 더 좋다. 하지만, querySelector는 구체적인 ID를 가지고 있지 않은 HTML 태그에 대해서도 접근 가능하다.

---

HTML Tag의 id를 통해 해당 원소를 반환한다. 해당 ID가 존재하지 않을때 NULL을 반환한다.
```js
element = document.getElementById(id);
```

selector의 구체적인 그룹과 일치하는 document 내의 첫번째 원소를 반환한다. 해당 selector가 존재하지 않을 때 NULL을 반환한다.
```js
element = document.querySelector(selectors);
```

예시)
```js
<form id="userForm">
  <input id="username" type="text" value="Guilherme" />
</form>
```
위 코드에 대해 username 요소를 얻고 싶고 이를 변수에 할당할때 각각의 경우는 아래와 같다.

- getElemenyById
```js
var username = document.getElementById("username");
```
- querySelector
```js
var username = document.querySelector("#userForm #username");
```

**Comment by yhshin0** [comment]
@yechoi42 
# 1. DOM
DOM은 HTML, XML 문서에 접근하기 위한 일종의 인터페이스이며, 계층화된 트리 구조로 이루어져 있습니다.

즉, HTML 문서를 파싱하여 나온 정보들을 노드 트리(node tree)라고 불리는 계층적 구조에 저장됩니다.

노드 트리는 최상위 레벨의 루트 노드부터 시작하여 가장 낮은 레벨의 텍스트 노드까지 뻗어 나갑니다.

![](http://tcpschool.com/lectures/img_js_htmldom.png)
출처: http://tcpschool.com/javascript/js_dom_concept

이렇게 계층화된 HTML 모델은 자바스크립트를 사용하여 다음과 같이 조작할 수 있습니다.
- 새로운 HTML 요소나 속성 추가.
- HTML 요소나 속성 제거.
- HTML 요소나 속성 변경.
- HTML 요소의 CSS 스타일 변경.
- HTML 요소에 HTML 이벤트 추가.

## 노드의 종류
HTML 문서를 구성하는 대표적인 노드의 종류는 다음과 같습니다.

|노드|설명|
|---|---|
|문서 노드(document node)|HTML 문서 전체를 나타내는 노드.|
|요소 노드(element node)|HTML 요소. 속성 노드를 가질 수 있는 유일한 노드.|
|속성 노드(attribute node)|HTML 요소의 속성. 요소 노드에 관한 정보를 가지고 있음(요소 노드의 자식 노드가 아님).|
|텍스트 노드(text node)|HTML 문서의 모든 텍스트.|
|주석 노드(comment node)|HTML 문서의 주석.|

## 노드 사이의 관계
노드 트리의 모든 노드는 서로 계층적 관계를 가지며, 다음과 같이 부모, 자식, 형제 노드로 이루어집니다.

![](http://tcpschool.com/lectures/img_js_node_relationship.png)
출처: http://tcpschool.com/javascript/js_dom_node

- 부모 노드는 자신을 갖고 있는 보다 한 단계 상위 노드를 의미합니다. 단 하나의 부모 노드만 가질 수 있습니다.
- 자식 노드는 자신이 갖고 있는 노드를 의미합니다.
- 형제 노드는 자신과 같은 부모 노드를 갖고 있는 동일한 레벨의 노드입니다.

노드 간에는 다음과 같은 프로퍼티로 정의되며, 해당 프로퍼티를 사용하여 접근할 수 있습니다.
- parentNode : 부모 노드
- childNodes : 자식 노드 리스트
- firstChild : 첫 번째 자식 노드
- lastChild : 마지막 자식 노드
- nextSibling : 다음 형제 노드
- previousSibling : 이전 형제 노드

# 2. HTML 요소 선택
자바스크립트에서는 HTML DOM을 사용하여 노드 트리의 모든 노드에 접근할 수 있습니다.

다음은 자주 사용되는 document 메소드입니다.

|메소드|설명|예제|
|---|---|---|
|document.getElementsByTagName(태그이름)|해당 태그 이름의 요소를 모두 선택함.|document.getElementsByTagName("div")|
|document.getElementById(아이디)|해당 아이디의 요소를 선택함.|document.getElementById("first")|
|document.getElementsByClassName(클래스이름)|해당 클래스에 속한 요소를 모두 선택함.|document.getElementsByClassName("btn")|
|document.getElementByName(name속성값)|해당 name 속성값을 가지는 요소를 모두 선택함.|document.getElementByName("age")|
|document.querySelector(선택자)|해당 선택자로 선택되는 첫번째 요소를 선택함.|document.querySelector("ul>li:nth-child(2)")|
|document.querySelectorAll(선택자)|해당 선택자로 선택되는 요소를 모두 선택함.|document.querySelectorAll("div.grade>li")|

# 3. EventListener
이벤트(event)란 웹 브라우저가 알려주는 HTML 요소에 대한 사건의 발생을 의미합니다.

document가 완전히 호출되었거나, HTML의 요소에 대해 마우스로 클릭을 하거나, 키보드의 특정 키를 누르는 등 이벤트가 발생했을 때 특정 객체 또는 요소에 이벤트 리스너를 등록할 수 있습니다.

이벤트 리스너를 등록하는 방법은 두가지 방법이 있습니다.
1. 이벤트의 대상이 되는 객체나 요소에 프로퍼티로 등록하는 방법
2. 객체나 요소의 메소드에 이벤트 리스너를 전달하는 방법

## 프로퍼티로 이벤트 리스너 등록
요소의 프로퍼티로 이벤트 리스너를 등록하려면 자바스크립트 코드에서 프로퍼티로 등록하거나, HTML 태그의 속성으로 등록하는 방법이 있습니다.

먼저 자바스크립트 코드를 사용하는 방법은 다음과 같은 방법이 있습니다.
```javascript
window.onload = function() {                    // 이 함수는 HTML 문서가 로드될 때 실행됨.
    var text = document.getElementById("text"); // 아이디가 "text"인 요소를 선택함.
    text.innerHTML = "HTML 문서가 로드되었습니다.";
}
```

다음과 같이 HTML 태그의 속성으로 등록할 수도 있습니다. 단, 이 방식은 코드의 가독성이 안 좋아지며, 유지보수가 힘들어집니다.
```html
<p onclick="alert('문자열을 클릭했어요!')">이 문자열을 클릭해 보세요!</p>
```

## 객체나 요소의 메소드에 이벤트 리스너를 전달하는 방법
자바스크립트의 addEventLister() 메소드를 사용하여 이벤트 리스너를 등록할 수 있습니다.
```javascript
대상객체.addEventListener(이벤트명, 실행할이벤트리스너, 이벤트전파방식)
```

addEventListener의 인자들은 다음과 같은 역할을 합니다.
1. 이벤트명 : 이벤트 리스너를 등록할 이벤트 타입을 문자열로 전달합니다.
2. 실행할 이벤트 리스너 : 지정된 이벤트가 발생했을 때 실행할 이벤트 리스너를 전달합니다.
3. 이벤트 전파 방식 : false면 버블링(bubbling) 방식으로, true면 캡처링(capturing) 방식으로 이벤트를 전파합니다.

다음은 예시입니다.
```javascript
var showBtn = document.getElementById("btn"); // 아이디가 "btn"인 요소를 선택함.
showBtn.addEventListener("click", showText);  // 선택한 요소에 click 이벤트 리스너를 등록함.
function showText() {
    document.getElementById("text").innerHTML = "짜잔~!! 텍스트가 나타났어요!!";
}
```

## 이벤트 리스너 삭제
이벤트 리스너를 삭제하려면 removeEventListener() 메소드를 사용하면 됩니다.
```javascript
function clickBtn() {
    btn.removeEventListener("mouseover", mouseoverBtn);
    btn.removeEventListener("mouseout", mouseoutBtn);
    document.getElementById("text").innerHTML = "이벤트 리스너가 삭제되었어요!";
}
```

---
출처: http://tcpschool.com/javascript/js_dom_concept
http://tcpschool.com/javascript/js_dom_document
http://tcpschool.com/javascript/js_dom_node
http://tcpschool.com/javascript/js_dom_nodeAccess
http://tcpschool.com/javascript/js_event_eventListenerRegister

**Reply by yechoi42** [reply]
와 넘 잘정리해주셨네요

**Comment by yechoi42** [comment]
@jwon42 
***
# DOM 요소의 선택
HTML 요소를 다루기 위해서는 우선 해당 요소를 선택해야만 합니다.

자바스크립트에서 특정 HTML 요소를 선택하는 방법은 다음과 같습니다.
- HTML 태그 이름으로
- id로
- class로
- name attribute(속성)dmfh
- css 선택자(selector)fh
- html 객체 집합(object collection)으로
 

## 1. HTML 태그 이름(tag name)을 이용한 선택

getElementsByTagName() 메소드는 HTML 태그 이름을 이용하여 HTML 요소를 선택합니다.

```
var selectedItem = document.getElementsByTagName("li"); // 모든 <li> 요소를 선택함.

for (var i = 0; i < selectedItem.length; i++) {

    selectedItem.item(i).style.color = "red"; // 선택된 모든 요소의 텍스트 색상을 변경함.

}
```


## 2. 아이디(id)를 이용한 선택

getElementById() 메소드는 아이디를 이용하여 HTML 요소를 선택합니다. 자바스크립트에서 아이디(id)를 이용한 선택은 해당 아이디를 가지고 있는 요소 중에서 첫 번째 요소 단 하나만을 선택합니다.

따라서 여러 요소를 선택하고 싶을 때는 태그 이름이나 클래스와 같은 다른 방법을 사용해야 합니다.

```
var selectedItem = document.getElementById("even"); // 아이디가 "even"인 요소를 선택함.

selectedItem.style.color = "red"; // 선택된 요소의 텍스트 색상을 변경함.
```


## 3. 클래스(class)를 이용한 선택

getElementsByClassName() 메소드는 클래스 이름을 이용하여 HTML 요소를 선택합니다.

```
var selectedItem = document.getElementsByClassName("odd"); // 클래스가 "odd"인 모든 요소를 선택함.

for (var i = 0; i < selectedItem.length; i++) {

    selectedItem.item(i).style.color = "red"; // 선택된 모든 요소의 텍스트 색상을 변경함.

}
```


## 4. name 속성(attribute)을 이용한 선택

getElementByName() 메소드는 HTML 요소의 name 속성(예를 들면 `<input name="checkbox"/>`에서 checkbox)을 이용하여 HTML 요소를 선택합니다.

```
var selectedItem = document.getElementsByName("first"); // name 속성값이 "first"인 모든 요소를 선택함.

for (var i = 0; i < selectedItem.length; i++) {

    selectedItem.item(i).style.color = "red"; // 선택된 모든 요소의 텍스트 색상을 변경함.

}
```

getElementById와의 차이점이라면 getElementById(아이디)로 가져오면 단일값을 가져오고 getElementsByName(이름)은 해당 이름으로 된 객체를 모두 배열로 받아올 수 있다는 점이다.


## 5. CSS 선택자(selector)를 이용한 선택

querySelectorAll() 메소드는 CSS 선택자(아이디, 클래스, 속성, 속성값 등)를 이용하여 HTML 요소를 선택합니다.

```
var selectedItem = document.querySelectorAll("li.odd"); // 클래스가 "odd"인 요소 중에서 <li> 요소만을 선택함.

for (var i = 0; i < selectedItem.length; i++) {

    selectedItem.item(i).style.color = "red"; // 선택된 모든 요소의 텍스트 색상을 변경함.

}
```

## 6. HTML 객체 집합(object collection)을 이용한 선택

HTML DOM에서 제공하는 객체 집합(object collection)을 이용하여 HTML 요소를 선택할 수 있습니다.

```
var title = document.title; // <title> 요소를 선택함.

document.write(title);
```

# DOM 조작시 유의할 점(Live Collection vs Static Collection)
```
<ul class="list">
    <li>1번</li>
    <li>2번</li>
    <li>3번</li>
</ul>
```
만약에 위 HTML코드에서 `<li>`태그를 자바스크립트를 이용하여 동적으로 제어하기 위해 돔을 가져온다고 해보자.

여러 가지 방법이 있겠지만 가장 흔하게 사용되는 방법은 아래와 같다.

```
const $li = document.querySelectorAll('.list li');

```
![image](https://user-images.githubusercontent.com/61399588/119262715-28389a80-bc17-11eb-9a21-26a37dada942.png)

콘솔 창에 출력해보면 원하던대로 li 요소들을 잘 가져오는 것을 확인할 수 있다.

또한 querySelectorAll()을 이용하여 가져 온 li 요소들은 NodeList라는 이름의 객체로 반환된다.

```
const $li = document.querySelector('.list').children;
```
![image](https://user-images.githubusercontent.com/61399588/119262731-3ab2d400-bc17-11eb-85c1-36fef738ceef.png)

이번에는 다른 방법으로 li 요소에 접근했는 데, 부모 요소인 ul의 children 속성을 사용하여 탐색하였다.

이 경우에도 아까와 똑같이 li 요소를 출력하였으나, 콘솔 창을 잘 보면 NodeList가 아닌 HTMLCollection이 반환된 것을 확인할 수 있다.

### NodeList
NodeList는 element.childNodes와 같은 속성이나 document.querySelectorAll()과 같은 메소드에 의해 반환되는 Node Collection이다. NodeList는 마치 Array와 비슷하게 생겼지만 배열과는 다르다. 우리는 배열과 같은 형태를 하고 있지만 배열이 아닌 객체들을 유사 배열이라고 부른다. 

NodeList는 배열이 아니기 때문에 배열에 사용 가능한 메소드를 대부분 사용할 수 없다. 그렇기 때문에 NodeList에 배열 메소드를 사용하기 위해서는 NodeList를 실제 배열로 변환하는 등의 방법을 사용해야 한다. 

(최신 브라우저에서는 NodeList에 forEach()와 같은 메소드를 지원하고 있다.)

중요한 점은 대개의 경우 NodeList는 라이브(Live) 콜렉션으로 DOM의 변경 사항을 실시간으로 반영한다. 그러나 document.querySelectorAll() 메소드에 의해 반환되는 NodeList는 정적(Static) 콜렉션으로 DOM의 변경 사항이 실시간으로 반영되지 않는다.


### HTMLCollection
HTMLCollection 인터페이스는 요소의 문서 내 순서대로 정렬된 일반 컬렉션을 나타내며 NodeList와 마찬가지로 유사 배열이다. HTMLCollection은 현대적인 DOM의 이전 세대부터 존재하던 구성요소로 element.children과 같은 속성 또는 

document.getElementsByClassName(), document.getElementsByTagName()과 같은 메소드에 의해 반환된다. HTMLCollection은 모두 문서가 바뀔 때 실시간으로 업데이트되는 라이브(Live) 콜렉션이다.



출처: 

http://tcpschool.com/javascript/js_dom_element
https://im-developer.tistory.com/110 [Code Playground]
https://link2me.tistory.com/649 [소소한 일상 및 업무TIP 다루기]

**Comment by jwon42** [comment]
## 1. DOM(문서객체모델)은 정확히 무엇일까?

DOM(Document Object Model)은 웹 페이지에 대한 인터페이스입니다. 기본적으로 여러 프로그램들이 페이지의 콘텐츠 및 구조, 그리고 스타일을 읽고 조작할 수 있도록 API를 제공합니다.



### 웹 페이지는 어떻게 만들어질까?

웹 브라우저가 원본 HTML 문서를 읽어들인 후, 스타일을 입히고 대화형 페이지로 만들어 뷰 포트에 표시하기까지의 과정을 “**Critical Rendering Path**”라고 합니다.이 단계들을 대략 두 단계로 나눌 수 있습니다. **첫 번째 단계에서 브라우저는 읽어들인 문서를 파싱하여 최종적으로 어떤 내용을 페이지에 렌더링할지 결정**합니다. 그리고 **두 번째 단계에서 브라우저는 해당 렌더링을 수행**합니다.

<img src="https://user-images.githubusercontent.com/61400214/119262947-fd027b00-bc17-11eb-9673-b4efc4c60671.jpg" width="500px">

첫 번째 과정을 거치면 “렌더 트리”가 생성됩니다.
렌더 트리는 웹 페이지에 표시될 HTML 요소들과 이와 관련된 스타일 요소들로 구성됩니다. 브라우저는 렌더 트리를 생성하기 위해 다음과 같이 두 모델이 필요합니다.

- DOM(Document Object Model) – HTML 요소들의 구조화된 표현
- CSSOM(Cascading Style Sheets Object Model) – 요소들과 연관된 스타일 정보의 구조화된 표현



### DOM은 어떻게 생성될까 (그리고 어떻게 보여질까)?

**DOM은 원본 HTML 문서의 객체 기반 표현 방식**입니다. 둘은 서로 비슷하지만, DOM이 갖고 있는 **근본적인 차이**는 **단순 텍스트로 구성된 HTML 문서의 내용과 구조가 객체 모델로 변환되어 다양한 프로그램에서 사용될 수 있다는 점**입니다.
DOM의 개체 구조는 “노드 트리”로 표현됩니다. 하나의 부모 줄기가 여러 개의 자식 나뭇가지를 갖고 있고, 또 각각의 나뭇가지는 잎들을 가질 수 있는 나무와 같은 구조로 이루어져 있기 때문입니다. 이 케이스의 경우, 루트 요소인 <html> 은 “부모 줄기”, 루트 요소에 내포된 태그들은 “자식 나뭇가지” 그리고 요소 안의 컨텐츠는 “잎”에 해당합니다.

아래 HTML 문서는

<img src="https://user-images.githubusercontent.com/61400214/119263215-18ba5100-bc19-11eb-80cd-fcf1730e3f21.jpg" width="500px">
이와 같은 노드 트리로 표현됩니다.
<img src="https://user-images.githubusercontent.com/61400214/119263223-1c4dd800-bc19-11eb-8864-4201cb7bce6b.jpg" width="500px">



### 주의할 점

위 예제 혹은 DevTools에서 DOM은 마치 HTML 문서와 1 대 1 매핑이 되는 것처럼 보입니다. 그러나 둘 간에는 몇 가지 차이점이 있습니다.
DOM이 무엇인지 완전히 이해하기 위해, 어떤 것이 DOM이 아닌지 살펴보겠습니다.



#### (1) DOM은 HTML이 아닙니다.

DOM은 HTML 문서로부터 생성되지만 항상 동일하지 않습니다. DOM이 원본 HTML 소스와 다를 수 있는 두 가지 케이스가 있습니다.

**작성된 HTML 문서가 유효하지 않을 때**
DOM은 유효한 HTML 문서의 인터페이스 입니다. DOM을 생성하는 동안, 브라우저는 유효하지 않은 HTML 코드를 올바르게 교정합니다.
아래 예시를 살펴보시면,
<img src="https://user-images.githubusercontent.com/61400214/119263263-515a2a80-bc19-11eb-900b-fff931a1e06f.jpg" width="500px">

문서에 유효한 HTML 규칙의 필수 사항인 `<head>` 와 `<body>` 요소가 빠져있습니다. 그렇지만 생성된 DOM 트리에는 올바르게 교정되어 나타납니다.
<img src="https://user-images.githubusercontent.com/61400214/119263267-56b77500-bc19-11eb-8c97-69a3337440ac.jpg" width="500px">

**자바스크립트에 의해 DOM이 수정될 때**
DOM은 HTML 문서의 내용을 볼 수 있는 인터페이스 역할을 하는 동시에 동적 자원이 되어 수정될 수 있습니다.
예를 들어, 자바스크립트를 사용해 DOM에 새로운 노드를 추가할 수 있습니다.
<img src="https://user-images.githubusercontent.com/61400214/119263298-7353ad00-bc19-11eb-9fe3-c7e07259daee.jpg" width="500px">
이 코드는 DOM을 업데이트합니다. 하지만 HTML 문서의 내용을 변경하진 않습니다.



#### (2) DOM은 브라우저에서 보이는 것이 아닙니다.

브라우저 뷰 포트에 보이는 것은 렌더 트리로 DOM과 CSSOM의 조합입니다. 렌더 트리는 오직 스크린에 그려지는 것으로 구성되어 있어 DOM과 다릅니다.
달리 말하면, 렌더링 되는 요소만이 관련 있기 때문에 시각적으로 보이지 않는 요소는 제외됩니다.
예를 들어, display: none 스타일 속성을 가지고 있는 요소입니다.
<img src="https://user-images.githubusercontent.com/61400214/119263314-7ea6d880-bc19-11eb-9629-2fec1676d675.jpg" width="500px">

DOM은 `<p>` 요소를 포함시킵니다.
<img src="https://user-images.githubusercontent.com/61400214/119263321-836b8c80-bc19-11eb-90cd-b6d8eb753032.jpg" width="500px">

그러나 렌더 트리에 해당하는 뷰 포트에 표시되는 내용은` <p>` 요소를 포함하지 않습니다.
<img src="https://user-images.githubusercontent.com/61400214/119263325-85cde680-bc19-11eb-9830-436cb6cccb55.jpg" width="500px">

#### (3) DOM은 개발도구에서 보이는 것이 아닙니다.

개발도구의 요소 검사기는 DOM과 가장 가까운 근사치를 제공합니다. 그러나 개발도구의 요소 검사기는 DOM에 없는 추가적인 정보를 포함합니다.
가장 좋은 예는 CSS의 가상 요소입니다. ::before 과 ::after 선택자를 사용하여 생성된 가상 요소는 CSSOM과 렌더 트리의 일부를 구성합니다.
하지만, 기술적으로 DOM의 일부는 아닙니다. DOM은 오직 원본 HTML 문서로부터 빌드 되고, 요소에 적용되는 스타일을 포함하지 않기 때문입니다.
가상 요소가 DOM의 일부가 아님에도 불구하고, 요소 검사기에서는 아래와 같이 확인됩니다.
<img src="https://user-images.githubusercontent.com/61400214/119263326-8a929a80-bc19-11eb-8e0a-e2916dfd010b.jpg" width="500px">



### 요약정리

DOM은 HTML 문서에 대한 인터페이스입니다. 첫째로 뷰 포트에 무엇을 렌더링 할지 결정하기 위해 사용되며,
둘째로는 페이지의 콘텐츠 및 구조, 그리고 스타일이 자바스크립트 프로그램에 의해 수정되기 위해 사용됩니다.
DOM은 원본 HTML 문서 형태와 비슷하지만 몇 가지 차이점이 있습니다.

- 항상 유효한 HTML 형식입니다.
- 자바스크립트에 수정될 수 있는 동적 모델이어야 합니다.
- 가상 요소를 포함하지 않습니다. (Ex. `::after`)
- 보이지 않는 요소를 포함합니다. (Ex. `display: none`)



------



**출처**

- DOM은 정확히 무엇일까? (번역)
  https://wit.nts-corp.com/2019/02/14/5522
- What, exactly, is the DOM? (원문)
  https://bitsofco.de/what-exactly-is-the-dom/
- 브라우저는 어떻게 동작하는가?
  https://d2.naver.com/helloworld/59361



------



**유용한 링크**

- 모던 JavaScript 튜토리얼
  https://ko.javascript.info/
- TCPschool 자바스크립트
  http://tcpschool.com/javascript/intro
